src/nodes/hypertable_modify.c

```diff
@@ -747,18 +747,22 @@ ExecModifyTable(CustomScanState *cs_node, PlanState *pstate)
 	 */
 	if ((operation == CMD_DELETE || operation == CMD_UPDATE) && !ht_state->comp_chunks_processed)
 	{
-		if (ts_cm_functions->decompress_target_segments)
+		/* Modify snapshot only if something got decompressed */
+		if (ts_cm_functions->decompress_target_segments &&
+			ts_cm_functions->decompress_target_segments(ht_state))
 		{
-			ts_cm_functions->decompress_target_segments(ht_state);
 			ht_state->comp_chunks_processed = true;
 			/*
 			 * save snapshot set during ExecutorStart(), since this is the same
 			 * snapshot used to SeqScan of uncompressed chunks
 			 */
 			ht_state->snapshot = estate->es_snapshot;
-			/* use current transaction snapshot */
-			estate->es_snapshot = GetTransactionSnapshot();
+
 			CommandCounterIncrement();
+			/* use a static copy of current transaction snapshot
+			 * this needs to be a copy so we don't read trigger updates
+			 */
+			estate->es_snapshot = RegisterSnapshot(GetTransactionSnapshot());
```

Comment:
Do you mean that on the next update cycle, we're trying the tuple previously modified by the BEFORE UPDATE trigger on the previous cycle? And the "invisible" error happens in `GetTupleForTrigger()`?

I think normally the new tuples from the previous update cycles are skipped because they are `TM_SelfModified` and not `TM_Invisible`. This is determined by `HeapTupleSatisfiesUpdate`. So if we're getting `TM_Invisible`, this means we either forgot to update the command counter after modifying the tuple, or we're taken the new snapshot with command counter higher than the previous update. I guess here it's the latter and you prevent that by making a copy of the snapshot?

Reply:
Normally the `es_snapshot` is initialized with a registered snapshot copy (`standard_ExecutorStart`), so this is probably correct. But I wonder if we should re-take the snapshot like this at all? Not sure if it's a standard practice. Normally the update code uses the `es_output_cid` as the current command id for locking the tuples, so maybe just incrementing the output cid after decompression would be enough?

Reply:
> But I wonder if we should re-take the snapshot like this at all? Not sure if it's a standard practice. Normally the update code uses the es_output_cid as the current command id for locking the tuples, so maybe just incrementing the output cid after decompression would be enough?

Looks like we must do it, otherwise we won't see the tuples we've decompressed. The `es_output_cid` comes into play later, when we do see the tuples and want to determine if they are updated by us.